test-utils/state-viewer/src/state_dump.rs

```diff
@@ -42,16 +42,77 @@ pub fn state_dump(
+        .iter()
+        .map(|(account_id, (public_key, amount))| AccountInfo {
+            account_id: account_id.clone(),
+            public_key: public_key.clone(),
+            amount: *amount,
+        })
+        .collect();
+    genesis_config.validators.sort();
+    // Record the protocol version of the latest block. Otherwise, the state
+    // dump ignores the fact that the nodes can be running a newer protocol
+    // version than the protocol version of the genesis.
+    genesis_config.protocol_version = last_block_header.latest_protocol_version();
+    let shard_config = runtime.get_shard_config(last_block_header.epoch_id()).unwrap();
+    genesis_config.shard_layout = shard_config.shard_layout;
+    genesis_config.num_block_producer_seats_per_shard =
+        shard_config.num_block_producer_seats_per_shard;
+    genesis_config.avg_hidden_validator_seats_per_shard =
+        shard_config.avg_hidden_validator_seats_per_shard;
+    // Record only the filename of the records file.
+    // Otherwise the absolute path is stored making it impossible to copy the dumped state to actually use it.
+    match records_path {
+        Some(records_path) => {
+            let mut records_path_dir = records_path.to_path_buf();
+            records_path_dir.pop();
+            fs::create_dir_all(&records_path_dir).unwrap_or_else(|_| {
+                panic!("Failed to create directory {}", records_path_dir.display())
+            });
+            let records_file = File::create(&records_path).unwrap();
+            let mut ser = serde_json::Serializer::new(records_file);
+            let mut seq = ser.serialize_seq(None).unwrap();
+            let total_supply =
+                get_records(runtime, state_roots, last_block_header, &validators, &mut |sr| {
+                    seq.serialize_element(&sr).unwrap()
+                });
+            seq.end().unwrap();
+            // `total_supply` is expected to change due to the natural processes of burning tokens and
+            // minting tokens every epoch.
+            genesis_config.total_supply = total_supply;
+            near_config.genesis =
+                Genesis::new_with_path(genesis_config, records_path.to_path_buf());
+            near_config.config.genesis_records_file =
+                Some(records_path.file_name().unwrap().to_str().unwrap().to_string());
+        }
+        None => {
+            let mut records: Vec<StateRecord> = vec![];
+            let total_supply =
+                get_records(runtime, state_roots, last_block_header, &validators, &mut |sr| {
+                    records.push(sr)
+                });
+            // `total_supply` is expected to change due to the natural processes of burning tokens and
+            // minting tokens every epoch.
+            genesis_config.total_supply = total_supply;
+            near_config.genesis = Genesis::new(genesis_config, records.into());
+        }
+    }
+    near_config
+}
 
+fn get_records(
```

Comment:
comment what this function does - (and consider renaming to something like 'iterate' - as it seems that you call the callback on each item rather than returning it)